### PR TITLE
Commercehub fixing bugs

### PR DESCRIPTION
**Pull request by Heavyblade**
## Summary:
Add changes to address some observations/improvements
in the CommerceHub certification.

* SER-494 / SER-498: setting and using order_id as reference
* SER-495: use the correct end-point for verify
* SER-496 / SER-497: get avs and cvv verification codes
* SER-500: getting an error code from response

## Remote Test:
Finished in 108.050361 seconds.
23 tests, 63 assertions, 0 failures, 0 errors,
0 pendings, 0 omissions, 0 notifications
100% passed

## Unit Tests:
Finished in 46.103032 seconds.
5458 tests, 77155 assertions, 0 failures, 0 errors,
0 pendings, 0 omissions, 0 notifications
100% passed

## RuboCop:
760 files inspected, no offenses detected

### CODE REVIEW

File: test/unit/gateways/commerce_hub_test.rb [hunk 1]
```diff
@@ -128,7 +128,7 @@ def test_failed_purchase_and_authorize
 
     response = @gateway.authorize(@amount, @credit_card, @options)
     assert_failure response
-    assert_equal 'HOST', response.error_code
+    assert_equal 'string', response.error_code
```
Comment (edgarv09) [inline test/unit/gateways/commerce_hub_test.rb:131]: just for curiosity, what exactly is being tested here? 
e.g. 
```
response.error_code == 'string' or response.error_code is a String 
```

**Review by edgarv09 — APPROVED** [review]
LGTM 🐰

File: test/unit/gateways/commerce_hub_test.rb (hunk 1 above)
Comment (Heavyblade) [reply test/unit/gateways/commerce_hub_test.rb:131]: Ahhh nothing special, that's the value of the response code on the string that is being used as a successful response:

tmux

File: lib/active_merchant/billing/gateways/commerce_hub.rb [hunk 2]
```diff
@@ -113,6 +116,9 @@ def add_transaction_interaction(post, options)
       def add_transaction_details(post, options, action = nil)
         post[:transactionDetails] = {}
         post[:transactionDetails][:captureFlag] = options[:capture_flag] unless options[:capture_flag].nil?
+        post[:transactionDetails][:merchantOrderId] = options[:order_id] if options[:order_id].present? && action == 'sale'
+        post[:transactionDetails][:merchantTransactionId] = options[:order_id] if options[:order_id].present? && action == 'sale'
+
```
Comment (gasb150) [inline lib/active_merchant/billing/gateways/commerce_hub.rb:121]: ❓ Why did you add to different fields the same value in the same condition?

💭 Also I think you could do something like this:

```ruby
if options[:order_id].present? && action == 'sale'
    post[:transactionDetails][:merchantOrderId] = options[:order_id]
    post[:transactionDetails][:merchantTransactionId] = options[:order_id]
```

File: lib/active_merchant/billing/gateways/commerce_hub.rb [hunk 3]
```diff
@@ -177,10 +183,12 @@ def build_purchase_and_auth_request(post, money, payment, options)
...
       def add_reference_transaction_details(post, authorization, options, action = nil)
         post[:referenceTransactionDetails] = {}
-        post[:referenceTransactionDetails][:referenceTransactionId] = authorization
+        post[:referenceTransactionDetails][:referenceTransactionId] = authorization unless authorization.match?(/^order_id/)
+
         if action != 'capture'
           post[:referenceTransactionDetails][:referenceTransactionType] = options[:reference_transaction_type] || 'CHARGES'
-          post[:referenceTransactionDetails][:referenceMerchantTransactionId] = options[:reference_merchant_transaction_id]
+          post[:referenceTransactionDetails][:referenceMerchantTransactionId] = authorization.match?(/^order_id/) ? authorization.split('=').last : options[:reference_merchant_transaction_id]
+          post[:referenceTransactionDetails][:referenceMerchantOrderId] = authorization.match?(/^order_id/) ? authorization.split('=').last : options[:reference_merchant_order_id]
```
Comment (gasb150) [inline lib/active_merchant/billing/gateways/commerce_hub.rb:191]: Same here.

**Review by gasb150 — COMMENTED** [review]
LGTM, I just left a few comments.

File: lib/active_merchant/billing/gateways/commerce_hub.rb (hunk 2 above)
Comment (Heavyblade) [reply lib/active_merchant/billing/gateways/commerce_hub.rb:121]: you are right 👍

File: lib/active_merchant/billing/gateways/commerce_hub.rb (hunk 3 above)
Comment (Heavyblade) [reply lib/active_merchant/billing/gateways/commerce_hub.rb:191]: going to refector to something like:

```ruby
          order_id = authorization.split('=').last if authorization.match?(/^order_id/) 
          post[:referenceTransactionDetails][:referenceMerchantTransactionId] =  order_id || options[:reference_merchant_transaction_id]
          post[:referenceTransactionDetails][:referenceMerchantOrderId] = order_id || options[:reference_merchant_order_id]

```

**Review by naashton — APPROVED** [review]
LGTM! Thanks for the quick work on this.
